app/src/main/java/net/squanchy/schedule/tracksfilter/ScheduleTracksFilter.kt

```diff
@@ -56,97 +63,133 @@ class ScheduleTracksFilterActivity : AppCompatActivity() {
+        val centerY = closeButton.y + closeButton.height / 2
+
+        ViewAnimationUtils.createCircularReveal(
+            filtersRoot,
+            centerX.toInt(),
+            centerY.toInt(),
+            hypot(filtersRoot.width.toDouble(), filtersRoot.height.toDouble()).toFloat(),
+            0F
+        ).apply {
+            duration = resources.getInteger(R.integer.track_filters_disappear_duration).toLong()
+
+            doOnEnd {
+                filtersRoot.visibility = View.INVISIBLE
+                finish()
+            }
+
+            start()
+        }
+    }
+
     override fun onStart() {
         super.onStart()
 
+        if (needsAppearAnimation.getAndSet(false)) {
+            prepareAppearanceAnimation()
+            filtersRoot.postOnAnimation { animateAppearing() }
+        }
+
         subscription = Observable.combineLatest(tracksRepository.tracks(), tracksFilter.selectedTracks, combineIntoCheckableTracks())
+            .filter { it.isNotEmpty() }
+            .distinctUntilChanged()
             .subscribeOn(Schedulers.computation())
             .observeOn(AndroidSchedulers.mainThread())
             .subscribe { checkableTracks ->
+                trackFiltersList.setAdapterIfNone(trackAdapter)
                 this.checkableTracks = checkableTracks
                 trackAdapter.submitList(checkableTracks)
             }
     }
 
-    private fun combineIntoCheckableTracks(): BiFunction<List<Track>, Set<Track>, List<CheckableTrack>> {
-        return BiFunction { tracks, selectedTracks ->
-            tracks.map { track -> CheckableTrack(track, selectedTracks.contains(track)) }
-        }
-    }
+    private val needsAppearAnimation: AtomicBoolean = AtomicBoolean(true)
+
+    @Suppress("MagicNumber") // Just animation code… needs a few magic numbers
+    private fun animateAppearing() {
+        val centerX = closeButton.x + closeButton.width / 2
+        val centerY = closeButton.y + closeButton.height / 2
+
+        ViewAnimationUtils.createCircularReveal(
+            filtersRoot,
+            centerX.toInt(),
+            centerY.toInt(),
+            0F,
+            hypot(filtersRoot.width.toDouble(), filtersRoot.height.toDouble()).toFloat()
+        ).apply {
+            val totalDuration = resources.getInteger(R.integer.track_filters_appear_duration).toLong()
+            val delay = resources.getInteger(R.integer.track_filters_appear_delay).toLong()
+
+            duration = totalDuration
+
+            filtersRoot.postOnAnimationDelayed(totalDuration / 2) {
+                dialogTitle.slideDownAndFadeIn(duration = totalDuration - delay, delay = 0)
+                dialogSubtitle.slideDownAndFadeIn(duration = totalDuration - 2 * delay, delay = delay)
+                trackFiltersList.slideDownAndFadeIn(duration = totalDuration - 3 * delay, delay = 2 * delay)
+            }
 
-    override fun onStop() {
-        super.onStop()
-        subscription?.dispose()
+            start()
```

Comment:
same here, I'd move `start` outside of `apply`

Reply:
Same as before. I don't get why.

Reply:
609e9a69c3e622999dd949801436f84ded7d768e